src/prng/hc128.rs

```diff
@@ -0,0 +1,543 @@
+            // Q block
+            self.t[cc+512+0]  = self.step_q(cc+0,  cc+1,  ee+13, ee+6,  ee+4);
+            self.t[cc+512+1]  = self.step_q(cc+1,  cc+2,  ee+14, ee+7,  ee+5);
+            self.t[cc+512+2]  = self.step_q(cc+2,  cc+3,  ee+15, ee+8,  ee+6);
+            self.t[cc+512+3]  = self.step_q(cc+3,  cc+4,  cc+0,  ee+9,  ee+7);
+            self.t[cc+512+4]  = self.step_q(cc+4,  cc+5,  cc+1,  ee+10, ee+8);
+            self.t[cc+512+5]  = self.step_q(cc+5,  cc+6,  cc+2,  ee+11, ee+9);
+            self.t[cc+512+6]  = self.step_q(cc+6,  cc+7,  cc+3,  ee+12, ee+10);
+            self.t[cc+512+7]  = self.step_q(cc+7,  cc+8,  cc+4,  ee+13, ee+11);
+            self.t[cc+512+8]  = self.step_q(cc+8,  cc+9,  cc+5,  ee+14, ee+12);
+            self.t[cc+512+9]  = self.step_q(cc+9,  cc+10, cc+6,  ee+15, ee+13);
+            self.t[cc+512+10] = self.step_q(cc+10, cc+11, cc+7,  cc+0,  ee+14);
+            self.t[cc+512+11] = self.step_q(cc+11, cc+12, cc+8,  cc+1,  ee+15);
+            self.t[cc+512+12] = self.step_q(cc+12, cc+13, cc+9,  cc+2,  cc+0);
+            self.t[cc+512+13] = self.step_q(cc+13, cc+14, cc+10, cc+3,  cc+1);
+            self.t[cc+512+14] = self.step_q(cc+14, cc+15, cc+11, cc+4,  cc+2);
+            self.t[cc+512+15] = self.step_q(cc+15, dd+0,  cc+12, cc+5,  cc+3);
+        }
+        self.counter1024 += 16;
+    }
+}
+
+impl Rng for Hc128Rng {
+    #[inline]
+    fn next_u32(&mut self) -> u32 {
+        if self.index >= 16 {
+            self.state.update(&mut self.results);
+            self.index = 0;
+        }
+
+        let value = self.results[self.index];
+        self.index += 1;
+        value
+    }
+
+    #[inline]
+    fn next_u64(&mut self) -> u64 {
+        let index = self.index;
+        if index < 15 {
+            self.index += 2;
+            // Read an u64 from the current index
+            if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
+                unsafe { *(&self.results[index] as *const u32 as *const u64) }
+            } else {
+                let x = self.results[index] as u64;
+                let y = self.results[index + 1] as u64;
+                (y << 32) | x
+            }
+        } else if index >= 16 {
+            self.state.update(&mut self.results);
+            self.index = 2;
+            let x = self.results[0] as u64;
+            let y = self.results[1] as u64;
+            (y << 32) | x
+        } else {
+            let x = self.results[15] as u64;
+            self.state.update(&mut self.results);
+            self.index = 1;
+            let y = self.results[0] as u64;
+            (y << 32) | x
+        }
+    }
+
+    // As an optimization we try to write directly into the output buffer.
+    // This is only enabled for platforms where unaligned writes are known to
+    // be safe and fast.
+    // This improves performance by about 12%.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
```

Comment:
Not entirely sure how I feel about this. Pretty cool that you can do this but a fair amount extra complexity. Well, we have it now!

Reply:
For ISAAC I would not think it worth it, but here it is directly filling at most 64 bytes at a time. It seems to me requesting that amount or more is not uncommon. So the 12% improvement seemed worth the complexity. It had the consequence that `Hc128` had to be a separate struct inside `Hc128Rng`. So the complexity cost is higher than that single function...

Reply:
The struct separation is good for readability anyway.